pkg/bgp/speaker/events.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2017 Google Inc.
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package speaker
+
+import (
+	"github.com/cilium/cilium/pkg/k8s"
+
+	"go.universe.tf/metallb/pkg/k8s/types"
+	metallbspr "go.universe.tf/metallb/pkg/speaker"
+)
+
+type svcEvent struct {
+	id  k8s.ServiceID
+	svc *metallbspr.Service
+	eps *metallbspr.Endpoints
+}
+type epEvent svcEvent
+
+// nodeEvent must be a pointer to a map because maps are not hashable, aka
+// cannot be keys in a map, which queue is.
+type nodeEvent *map[string]string
+
+// run runs the reconciliation loop, fetching events off of the queue to
+// process. The events supported are svcEvent, epEvent, and nodeEvent. This
+// loop is only stopped (implicitly) when the Agent is shutting down.
+//
+// Adapted from go.universe.tf/metallb/pkg/k8s/k8s.go.
+func (s *Speaker) run() {
```

Comment:
Oh, and I think there's commonality here between this logic, the core of `pkg/serializer`, and the core of `pkg/eventqueue` where I think we have the opportunity to combine & improve one implementation of a serialized queue that handles concurrent queueing, provides introspection capabilities, etc. I don't think it's reasonable to make any such change right now but it might be worth familiarizing yourself with those packages as well to improve on the code in future.

Reply:
Ah, good to know. Upon quick inspection of `pkg/eventqueue`, it seems that would actually a fit for this code. I'll add that as a followup. Thanks